### ISSUE TITLE
Decouple download and generate scripts

### ISSUE THREAD

**Issue by behnam** [behnam/rust-unic#140]
The requirements we have for download and generate scripts are fairly different.

For download, we want to keep files from every *source* in sync, as we read the version from their `ReadMe.txt`. So, for each source, we want to delete all the files before re-downloading. The files under `/data/`, for each source, should never go out of sync, version-wise. Because of this, this part of the code should be grouped together based on the data *source*.

For generate, in contrast, we want to have steps the generate consistent table files, which depend on one or more sources, and generates one or more table files. For this part of the code, it makes sense to group them based on *major component*, because that's closer to the development process.

The current model that has coupled download and generate code to be called from one binary makes it hard to satisfy this pattern.

Also, with the current model, we cannot add the download logic for a new source, like `emoji`, in one step and get to the generate code later; forcing us to put in a dummy generate code until we get to the implementation.

Also, the two steps don't share any model configuration, and only sharing some CLI implementation.

I think the current model doesn't have much benefits and it's going to be much easier if we have two separate creates for these tasks, each with their own functionality.

Also, now we can rename `gen` to `tools`, as we have killed all the Python code. So, we can have `/tools/download/` and `/tools/generate/`, each a separate binary create. Or, we can have one `tools` create with multiple binaries inside.

**Comment by behnam** [comment]
What do you think about this, @CAD97?

**Comment by CAD97** [comment]
I still think it's very useful to have a single command to both download and generate all of the necessary things. But that can be handled by a shell script, I suppose.

I'm fully open to decoupling the two. They'll always be somewhat coupled as `generate` relies on `download` having run and agreeing on where files are put, but it does make sense to split the two so they can have separate concepts of "data source" versus "component destination".

Maybe then we can actually get the cached-compile benefit for the download portion (the reason there's a config read in from the filesystem rather than just being hardcoded or `import!`ed) even while adding new output to the generate portion.

I'm actually poking around in it currently to add parsing for `PropList.txt` and `DerivedCoreProperties.txt`, which are both big shared files including many binary properties. Once I'm done with that, I can do this this weekend.

TL;DR I had my reasons for sticking them in one binary when building it -- mainly, being lazy -- but see only positives for splitting them now.

**Comment by behnam** [comment]
Yeah, I think we started with the hope to share some data between them, like the data file paths. But, in practice, the path is hard-coded into an `include_str!` for the *generate* portion, while it's in a config file for the *download* portion, which demands more complexities just to share a few strings.

I'm looking into simplifying some parts of the `download` module right now, making it be more simple, less verbose. I'll let you know how it goes, and you can work on splitting `gen` whenever you're done. Thanks!

About the new props, that's great! I just added `PropList.txt` to the download list, because I wanted to write the code for `Bidi_Control`. Now that we have the `BinaryCharProperty` model, we can add most of the other props. How far are you in the implementations?

Looking at the code for `Bidi_Mirrored` to implement `Bidi_Control`, I noticed that most of the code can be put into a macro. Basically, there are just a bunch of names, and the rest is straightforward. Are you doing this already?

If not, we can just add what we need right now, and then use the macro for the rest of them next. What do you think?

**Comment by CAD97** [comment]
I'm only working on the pulling of the data files into the generate module for further use. It's done for DerivedCoreProperties, and adding PropList is just a matter of an unfortunate amount of boilerplate to actually load the data into a structure easily used.

Download can almost certainly be simplified by using something other than Hyper. I chose it mainly because it was the one API I already roughly knew how to use. Feel free to change it if that will simplify everything though.

I haven't worked on an extention to the `char_property!` macro for binary properties yet, but now that `BinaryCharProperty` is in it's on my list. I'd say implement `Bidi_Control` by hand, like was done for the other properties initially, and then once we have the small number of manual binary properties a macro can be made to make adding more (and there are a lot more binary properties, oh boy) much easier.

**Comment by CAD97** [comment]
See https://github.com/CAD97/rust-unic/commit/be84e1dccafaaed8255abec905d8cb5513a5e8d1 for DerivedCoreProperties. Adding PropList currently, shouldn't take _too_ long.

**Comment by behnam** [comment]
Okay, @CAD97, <https://github.com/behnam/rust-unic/pull/145> simplifies the `download` module and makes it ready to become a separate crate/binary. I'll leave the rest of this issue to you, as you've spent a lot more time on the `generate` code.

In the meanwhile, we can easily add new files to the download config (and just download), and the refresh-data script that calls into the coupled binary would work file with `normal` being passed in.

**Comment by behnam** [comment]
@CAD97, I'm ended up working on the generate scripts and there's a lot of changes on my tree now. Please let me finish with that work, and we review, and we can decide about this issue then.

Here's the tree: https://github.com/behnam/rust-unic/tree/tables

**Comment by behnam** [comment]
Okay, now we have three major modules under `unic-gen`:
* `download`, which downloads the source data files based on a config file.
* `source`, which provides parsers and data models for the source data files.
* `writer`, which is responsible for generation of data table files for UNIC components, using `source` and its own internal intermediate data models.

Now, the only connection between `download` and the other two is the local file paths for the data files. In `download`, these paths are auto-generated from the config file, and in `source`, they are hardcoded in strings.

IMHO, we don't need further tangling between `download` and `source`, and it's fine to just make an assumption that the file needed is there. (Otherwise the script to generate data tables will fail with error for us, the developers.) Therefore, I think it's fine to just have two separate binaries for these two areas.

Right now, the `main.rs` file has some CLI code in common for the two areas. I think it's fine to just have copies of CLI code, since it's not that large.

Since the current code is just working fine, it's not a priority. We can leave this task open for anyone interested to grab. It's a good started task to learn more about CLI in Rust, and the data generation in UNIC.
